src/software/amazon/ion/Timestamp.java

```diff
@@ -2051,119 +1921,144 @@ private static void print_fractional_digits(Appendable out, BigDecimal value)
     //=========================================================================
     // Timestamp arithmetic
 
-
     /**
      * Returns a timestamp relative to this one by the given number of
-     * milliseconds.
+     * milliseconds. Uses this Timestamp's configured Calendar to perform the
+     * arithmetic.
      *
      * @param amount a number of milliseconds.
      */
-    public final Timestamp addMillis(long amount)
-    {
-        if (amount == 0) return this;
-
-        // This strips off the local offset, expressing our fields as if they
-        // were UTC.
-        BigDecimal millis = make_localtime().getDecimalMillis();
-        millis = millis.add(BigDecimal.valueOf(amount));
-
-        Timestamp ts = new Timestamp(millis, _precision, _offset);
-
-        // Anything with courser-than-millis precision will have been extended
-        // to 3 decimal places due to use of getDecimalMillis().  Fix that.
-        ts._fraction = _fraction;
-        if (_offset != null && _offset != 0)
-        {
-            ts.apply_offset(_offset);
+    public final Timestamp addMillis(long amount) {
+        if (amount == 0 && _precision == Precision.SECOND && _fraction != null && _fraction.scale() >= 3) {
+            // Zero milliseconds are to be added, and the precision does not need to be increased.
+            return this;
         }
+        long seconds = amount / 1000;
+        BigDecimal millis = BigDecimal.valueOf(amount % 1000).movePointLeft(3);
+        if (_fraction != null) {
+            millis = _fraction.add(millis);
+        }
+        BigDecimal newFraction;
+        if (BigDecimal.ONE.compareTo(millis) <= 0) {
+            newFraction = millis.subtract(BigDecimal.ONE);
+            seconds += 1;
+        } else if (BigDecimal.ZERO.compareTo(millis) > 0) {
+            newFraction = BigDecimal.ONE.add(millis);
+            seconds -= 1;
+        } else {
+            newFraction = millis;
+        }
+        Timestamp ts = this;
+        do {
```

Comment:
suggest encapsulating this loop into a `private Timestamp addSecond(long seconds)` method

Reply:
Done.